functions/access.js

```diff
@@ -52,65 +60,95 @@ const getAccessDoc = async (venueId, method) => {
 };
 
 const isValidPassword = async (venueId, password) => {
+  if (!venueId || !password) return false;
+
   const access = await getAccessDoc(venueId, "password");
-  if (!access.exists || !access.data().password) {
+
+  if (!access || !access.exists || !access.data().password) {
     return false;
   }
+
   return passwordsMatch(access.data().password, password);
 };
 
 const isValidEmail = async (venueId, email) => {
+  if (!venueId || !email) return false;
+
   const access = await getAccessDoc(venueId, "emails");
-  if (!access.exists || !access.data().emails) {
+
+  if (!access || !access.exists || !access.data().emails) {
     return false;
   }
+
   return access.data().emails.includes(email.trim().toLowerCase());
 };
 
 const isValidCode = async (venueId, code) => {
+  if (!venueId || !code) return false;
+
   const access = getAccessDoc(venueId, "code");
-  if (!access.exists || !access.data().codes) {
+
+  if (!access || !access.exists || !access.data().codes) {
     return false;
   }
+
   return access.data().codes.includes(code.trim());
 };
 
 const createToken = async (venueId, uid, password, email, code) => {
-  if (!venueId || !uid || !password || !email || !code) return {};
+  if (!venueId || !uid || !password || !email || !code) return undefined;
 
-  const venue = await admin.firestore().collection("venues").doc(venueId).get();
-  if (!venue.exists) {
+  const venueRef = admin.firestore().collection("venues").doc(venueId);
+  const accessRef = admin.firestore().collection("accessgranted").doc(uid);
+
+  if (!venueRef.exists) {
     throw new HttpsError("not-found", `venue ${venueId} does not exist`);
   }
```

Comment:
Does `.exists` work on a ref that hasn't been `.get()` yet?

Reply:
It does not. I fixed this.

Reply:
Sweet, thanks :)